NOISSUE - Add property based testing to users service

rodneyosodo opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

This is a feature PR as it adds property-based testing to users API. In addition it is a refactor to error handling on the API
<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: #1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

Adds [schemathesis](https://github.com/schemathesis/schemathesis#getting-started), property-based testing to users service
<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves #1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

No issue

## Have you included tests for your changes?

Yes, I have included tests for my changes.

<!--If you have not included tests, please explain why.
For example:
Yes, I have included tests for my changes.
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

No
<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

rodneyosodo committed:
Add property based testing to users API using schemathesis

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

Stranger6667 commented on .github/workflows/api-tests.yml:
```diff
@@ -0,0 +1,54 @@
+          base-url: http://localhost:9002
+          checks: all
+          report: false
+          args: '--header "Authorization: Bearer ${{ env.USER_TOKEN }}" --contrib-unique-data --contrib-openapi-formats-uuid --hypothesis-suppress-health-check=filter_too_much --stateful=links'
```
The `--contrib-unique-data` option is deprecated and going to be removed in the next major version of Schemathesis. I suggest avoiding it as it leads to unstable tests

drasko replied:
@rodneyosodo can you please inspect this and see if we can remove `--contrib-unique-data` and what implications it will make (i.e. how we can find a future-proof approach)

rodneyosodo committed:
Refactor API test arguments to remove deprecated options

Removed the `--contrib-unique-data` option from API test arguments in `.github/workflows/api-tests.yml` and `Makefile` since it will be deprecated

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

arvindh123 commented on api/openapi/users.yml:
```diff
@@ -305,15 +362,20 @@ paths:
 
   /password/reset-request:
     post:
+      operationId: requestPasswordReset
```
`/password` is comes under the `/users` route in present `main` branch

arvindh123 commented on users/api/clients.go:
```diff
@@ -139,6 +126,22 @@ func clientsHandler(svc users.Service, r *chi.Mux, logger *slog.Logger, pr *rege
 		), "disable_client").ServeHTTP)
 	})
 
+	r.Route("/password", func(r chi.Router) {
+		r.Post("/reset-request", otelhttp.NewHandler(kithttp.NewServer(
+			passwordResetRequestEndpoint(svc),
+			decodePasswordResetRequest,
+			api.EncodeResponse,
+			opts...,
+		), "password_reset_req").ServeHTTP)
+
+		r.Put("/reset", otelhttp.NewHandler(kithttp.NewServer(
+			passwordResetEndpoint(svc),
+			decodePasswordReset,
+			api.EncodeResponse,
+			opts...,
+		), "password_reset").ServeHTTP)
+	})
+
```
Why we are going to change these routes for password?

rodneyosodo replied:
From SDK it is under` /password` that is why I made the change to API docs and code base https://github.com/absmach/magistrala/blob/364749f28182d4e4a7a317e605f38c5c2b6aa222/pkg/sdk/go/users.go#L204

rodneyosodo replied:
We agreed and reverted to using non users prefix on password endpoint

arvindh123 reviewed (changes requested):
LGTM 

At [here](https://github.com/rodneyosodo/magistrala/blob/5e139ad19603ac4efa3d4b9d3f2bc148a8184683/users/postgres/clients.go#L78-L94) in users repository , we are using svcerr, instead of svcerr, does the repoerr will fit there ?

arvindh123 commented on Makefile:
```diff
@@ -129,6 +129,16 @@ test: mocks
 	go test -v --race -count 1 -tags test -coverprofile=coverage/coverage.out $$(go list ./... | grep -v 'consumers\|readers\|postgres\|internal\|opcua\|cmd')
 
+test_api:
+	@which st > /dev/null || (echo "schemathesis not found, please install it from https://github.com/schemathesis/schemathesis#getting-started" && exit 1)
+	st run api/openapi/users.yml \
+	--checks all \
+	--base-url http://localhost:9002 \
+	--header "Authorization: Bearer $(USER_TOKEN)" \
+	--contrib-openapi-formats-uuid \
+	--hypothesis-suppress-health-check=filter_too_much \
+	--stateful=links
```
It will be better if we impove the this command, like If USER_TOKEN is empty throw error or add mandatory required arguments username, password and domain to this command to get user_token

rodneyosodo replied:
Updated with [`0f27a04` (#2087)](https://github.com/absmach/magistrala/pull/2087/commits/0f27a04f67d4f2202a948a7c8748c13fa67ffef7)

rodneyosodo committed:
check USER_TOKEN is not empty

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
remove svcerr errors from repository layer

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
add domain login

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

arvindh123 commented on internal/groups/service.go:
```diff
@@ -696,7 +694,7 @@ func (svc service) changeGroupStatus(ctx context.Context, token string, group gr
 	}
 	if dbGroup.Status == group.Status {
-		return groups.Group{}, mgclients.ErrStatusAlreadyAssigned
+		return groups.Group{}, errors.ErrStatusAlreadyAssigned
```
@rodneyosodo, No need to change the following comments in this PR. 

We need to have some kind of error layers. 

Idea is to prevent or avoid sending of Reposistory errors in API response 

One of the way which I have in my mind to implement this idea is, reposistory should always return erorrs from pacakge `repoerr`
service should return errors from package `svcerr`, 

In API error encoder, We need to check for api errors and service errors and send only api errors and service errors in response and don't want to send repo errors in response. 

Example: 
for GET request of `things/<thingID>`
In response If entity is not found, then i should get error message of `svcerr.ErrNotFound`, not from repository layer, error return by postgres library.

In logs all the errors should be present. So if there we can get the more error details from service logs.

rodneyosodo replied:
This can be done with something like this https://github.com/absmach/magistrala-ui/pull/240 whereby all errors from the repository layer must be wrapped with the service layer unless specified otherwise. But it is highly intrusive.

arvindh123 replied:
Created a issue [here](https://github.com/absmach/magistrala/issues/2117) for this comment
